### PR TITLE
Bug fixes & code improvements

### PR DESCRIPTION
**Pull request by yacinehmito**
I am sharing the couple of bug fixes I needed to make for writing my memoir.
See the commit messages for the details.

### CODE REVIEW

**Comment by jxxcarlson** [comment]
Am in Morocco on vacation — will look at this at the end of the week when I return.

Thanks!

**Comment by gpyh** [comment]
If you have any question, feel free to ask.

File: lib/asciidoctor/latex/environment_block.rb [hunk 1]
```diff
@@ -92,8 +92,7 @@ class EnvironmentBlock < Asciidoctor::Extensions::BlockProcessor
 
     named :env
     on_context :open
-    # parse_context_as :complex
-    # ^^^ The above line gave me an error.  I'm not sure what do to with it.
+    parse_content_as :compound
```
Comment (mojavelinux) [inline lib/asciidoctor/latex/environment_block.rb]: This is correct. It's also the default. But it's good to be explicit.

File: lib/asciidoctor/latex/environment_block.rb [hunk 2]
```diff
@@ -153,7 +152,7 @@ def process parent, reader, attrs
         warn "for rode = code, attrs = #{attrs}".cyan
         block = create_block parent, :listing, reader.lines, attrs
       else
-        block = create_block parent, :environment, reader.lines, attrs
+        block = create_block parent, :environment, reader.lines, attrs, content_model: :compound
```
Comment (mojavelinux) [inline lib/asciidoctor/latex/environment_block.rb]: This is correct. The default content model is `:simple`. By making it `:compound`, you allow for nested blocks.

File: lib/asciidoctor/latex/node_processors.rb [hunk 3]
```diff
@@ -653,6 +641,10 @@ def image_process
         width = '2.5truein'
       end
       raw_image = self.attributes['target']
+      imagesdir = "#{document.attributes['imagesdir']}"
+      unless imagesdir.empty?
```
Comment (mojavelinux) [inline lib/asciidoctor/latex/node_processors.rb]: Alternately, you can write this as:

```ruby
unless (imagesdir = document.attr 'imagesdir').nil_or_empty?
  raw_image = ::File.join imagesdir, raw_image
end
```

Every object has a `nil_or_empty?` method in an Asciidoctor environment.

**Comment by mojavelinux** [comment]
I'm very glad to see the addition of the braces method to handle an arbitrary number of arguments. It's always good to avoid handling only a fixed number of positional arguments.

**Comment by gpyh** [comment]
Thanks for the review. I amended the commits to add your snippet about `imagesdir` and clean up the code in the *Fix problematic empty line in split* commit.

**Comment by gpyh** [comment]
What is the status on this?

**Comment by gpyh** [comment]
Bumping.

**Comment by prudhomm** [comment]
has this branch been merged into master?

**Comment by gpyh** [comment]
It hasn't been.

**Comment by yacinehmito** [comment]
In light of the recent activity on the repo, I'm bumping this PR.

**Comment by tribut** [comment]
Not sure if it means anything to you guys, but I reviewed at all the commits in this PR and recommend merging.

It fixes bugs, does some necessary cleanup and adds support for the toc macro.

**Comment by mojavelinux** [comment]
@yacinehmito Can you add a new commit (or force push a previous one) so that the CI job runs?

**Comment by mojavelinux** [comment]
> Not sure if it means anything to you guys, but I reviewed at all the commits in this PR and recommend merging.

Of course that means something! It means a lot, in fact. Thanks!

**Comment by jxxcarlson** [comment]
@yacinehmito @mojavelinux @tribut @prudhomm Thank you all!

**Comment by jxxcarlson** [comment]
Oops, there are some failing tests -- likely the tests are at fault.  They are very sensitive to minor changes.  I will try to get this resolved in the next few hours.

**Comment by jirutka** [comment]
IIRC you hadn’t finished the tests and many of them were failing since ever, but it’s quite a long time, so I’m not sure.

Doctest examples in TeX are indeed sensitive to any kind of change in output, ’cause there’s no normalization. I still don’t know if it’s even possible to write generic normalization and/or pretty-formatter for TeX syntax.

HTML examples are not sensible for changes that do not have any semantic meaning (like order of attributes, formatting etc.), because Doctest compares them in a normalized form.

**Comment by yacinehmito** [comment]
Thanks!

**Pull request merged**